AMDGPU/GlobalISel: add RegBankLegalize rules for bitfield extract

petar-avramovic opened the pull request:
Divergent S32 instruction is available, for S64 need to lower to S32.
Uniform instructions available for both S32 and S64 but need to pack
bitfield offset and size of bitfield into S32. Uniform instruction is
straight up selected since there is no available isel pattern.

petar-avramovic commented:
* **#132385** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132385?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132384** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132384?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132383** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132383?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132382** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132382?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#132381** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132381?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/132381?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -225,6 +228,103 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
     MI.eraseFromParent();
     break;
   }
+  case Div_BFE: {
```
I feel like the contents of the `case` should be a helper, otherwise the switch becomes hard to read.
The code is also easier to follow with less indentation, and you can add a comment on the function to explain why/how it's lowered like this

nhaehnle replied:
Yes, that makes sense.

Actually, this could be a motivation for being able to directly specify a helper function to be used as part of the legalizer rule definition...

petar-avramovic replied:
interesting, will look into that

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -225,6 +228,103 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
+    Register Src = MI.getOperand(FirstOpnd).getReg();
+    Register LSBit = MI.getOperand(FirstOpnd + 1).getReg();
+    Register Width = MI.getOperand(FirstOpnd + 2).getReg();
+    // Comments are for signed bitfield extract, similar for unsigned. x is sign
```
IMO these comments are confusing, what do other reviewers think?

nhaehnle replied:
I think they're helpful.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -225,6 +228,103 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
+  case Div_BFE: {
+    Register Dst = MI.getOperand(0).getReg();
+    assert(MRI.getType(Dst) == LLT::scalar(64));
+    bool Signed = isa<GIntrinsic>(MI) ? MI.getOperand(1).getIntrinsicID() ==
+                                            Intrinsic::amdgcn_sbfe
+                                      : MI.getOpcode() == AMDGPU::G_SBFX;
```
Move this into a `isSignedBFE` helper and reuse it below too ?
A 3 line ternary is a bit hard to read IMO

nhaehnle replied:
Considering that the next line also inspects the exact type of MI, I'm not convinced that this would make things cleaner.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -225,6 +228,103 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
+    // Comments are for signed bitfield extract, similar for unsigned. x is sign
+    // bit. s is sign, l is LSB and y are remaining bits of bitfield to extract.
+
+    // Src >> LSBit Hi|Lo: x?????syyyyyyl??? -> xxxx?????syyyyyyl
+    unsigned SHROpc = Signed ? AMDGPU::G_ASHR : AMDGPU::G_LSHR;
+    auto SHRSrc = B.buildInstr(SHROpc, {{VgprRB, S64}}, {Src, LSBit});
+
+    auto ConstWidth = getIConstantVRegValWithLookThrough(Width, MRI);
+
+    // Expand to Src >> LSBit << (64 - Width) >> (64 - Width)
+    // << (64 - Width): Hi|Lo: xxxx?????syyyyyyl -> syyyyyyl000000000
+    // >> (64 - Width): Hi|Lo: syyyyyyl000000000 -> ssssssssssyyyyyyl
+    if (!ConstWidth) {
+      auto Amt = B.buildSub(VgprRB_S32, B.buildConstant(SgprRB_S32, 64), Width);
+      auto SignBit = B.buildShl({VgprRB, S64}, SHRSrc, Amt);
+      B.buildInstr(SHROpc, {Dst}, {SignBit, Amt});
+      MI.eraseFromParent();
+      return;
+    }
```
Why do you do shr -> shl -> shr? It's cheaper to just do shl -> shr with appropriate widths.

Another issue here: What if only the data source operand is divergent, and the other inputs are uniform? I feel like we should handle that appropriately.

petar-avramovic replied:
Did not give it too much thought, just copied original implementation.
Need 4 instructions either way, instructions are two shifts and sub, fourth can be either sub or shift. Shift is faster then sub?
```
SHRSrc = Src >> LSBit
Amt = 64 - Width
SignBit = SHRSrc << Amt
Res = SignBit >> Amt
```

```
AmtR = 64 - Width
AmtL = AmtR - LSBit
SignBit = Src << AmtL
Res = SignBit >> AmtR
```

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,117 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
+  auto FieldOffset = B.buildAnd(SgprRB_S32, LSBit, Mask);
+  auto Size = B.buildShl(SgprRB_S32, Width, B.buildConstant(SgprRB_S32, 16));
+  auto Src1 = B.buildOr(SgprRB_S32, FieldOffset, Size);
+  unsigned Opc32 = Signed ? AMDGPU::S_BFE_I32 : AMDGPU::S_BFE_U32;
+  unsigned Opc64 = Signed ? AMDGPU::S_BFE_I64 : AMDGPU::S_BFE_U64;
```
There are no tablegen patterns for S_BFE, known bits for GlobalISel are missing support for many generic opcodes not to mention target intrinsics. Can we leave that for another patch?

arsenm replied:
> There are no tablegen patterns for S_BFE

Since the two operands are merged into one operand, I think you need to manually select it. Should add a fixme to stop directly selecting it

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,117 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
+  // Expand to Src >> LSBit << (64 - Width) >> (64 - Width)
+  // << (64 - Width): Hi|Lo: xxxx?????syyyyyyl -> syyyyyyl000000000
+  // >> (64 - Width): Hi|Lo: syyyyyyl000000000 -> ssssssssssyyyyyyl
+  if (!ConstWidth) {
```
I assume so, looks like intrinsics were introduced first, then the G_ opcodes

petar-avramovic commented:
ping

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,120 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
   MI.eraseFromParent();
 }
 
+bool isSignedBFE(MachineInstr &MI) {
```
```suggestion
static bool isSignedBFE(const MachineInstr &MI) {
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,120 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
+    switch (MI.getOperand(1).getIntrinsicID()) {
+    case Intrinsic::amdgcn_sbfe:
+      return true;
+    case Intrinsic::amdgcn_ubfe:
+      return false;
```
Don't really need to handle the false cases. Just the two positives

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,120 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
+  }
+}
+
+void RegBankLegalizeHelper::lowerDiv_BFE(MachineInstr &MI) {
```
"Div" may be confused with division. This also isn't really about divergence, but VGPR. lowerV_BFE?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.h:
```diff
@@ -168,6 +168,8 @@ enum LoweringMethodID {
   DoNotLower,
   VccExtToSel,
   UniExtToSel,
+  Uni_BFE,
+  Div_BFE,
```
V vs. S, not uniform vs scalar

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,117 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
+  // field offset in low and size in high 16 bits.
+
+  // Src1 Hi16|Lo16 = Size|FieldOffset
+  auto Mask = B.buildConstant(SgprRB_S32, maskTrailingOnes<unsigned>(6));
+  auto FieldOffset = B.buildAnd(SgprRB_S32, LSBit, Mask);
```
buildZextInReg builds multiple instructions and does not set reg banks, this way we set all register banks

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -127,6 +131,112 @@ void RegBankLegalizeHelper::widenLoad(MachineInstr &MI, LLT WideTy,
 }
 
+static bool isSignedBFE(MachineInstr &MI) {
+  if (isa<GIntrinsic>(MI)) {
```
nit: How about dyn_cast + use the `.is(IID)` helper?

Pierre-vh reviewed (approved):
LGTM but please wait in case @arsenm has more comments

petar-avramovic committed:
AMDGPU/GlobalISel: add RegBankLegalize rules for bitfield extract

Divergent S32 instruction is available, for S64 need to lower to S32.
Uniform instructions available for both S32 and S64 but need to pack
bitfield offset and size of bitfield into S32. Uniform instruction is
straight up selected since there is no available isel pattern.
